### PR TITLE
More Position info and Traversals

### PR DESCRIPTION
**Pull request by kRITZCREEK**
This is a byproduct of me playing around with the AST in: https://github.com/kRITZCREEK/morgana

I just wanted to confirm with you that we could make these kind of changes without breaking anything.

### CODE REVIEW

File: src/Language/PureScript/Parser/Declarations.hs [hunk 1]
```diff
@@ -531,25 +533,27 @@ parseIdentifierAndBinder =
...
     PositionedBinder
     ( P.buildExpressionParser operators
     . buildPostfixParser postfixTable
     $ parseBinderAtom
     )
   where
-  operators =
-    [ [ P.Infix (P.try (indented *> parseOpBinder P.<?> "binder operator") >>= \op ->
-          return (BinaryNoParensBinder op)) P.AssocRight
+    flattenPositionedBinder (PositionedBinder p c (PositionedBinder p' _ b)) | p == p' = PositionedBinder p c b
```
Comment (paf31) [inline src/Language/PureScript/Parser/Declarations.hs]: It seems like this shouldn't be necessary.

**Comment by paf31** [comment]
This seems fine to me, but we would need to go through all of the traversals and recursive functions on these types to make sure they don't need to be updated. I suspect a few would. We might be better off waiting for @garyb's work to add annotations, and then we can move position info into the annotations.

The traversal changes look necessary though, so good catch on that :)

**Comment by kRITZCREEK** [comment]
Okay, I'll remove the changes to the parser soon then.

We're still missing a couple of traversal cases for `BoundValueDeclaration`, but I can't figure all of them out. Maybe we could do them together? I'd like to understand what's going on in some of these Traversals :D

**Comment by paf31** [comment]
Sure, ping me on IRC and we'll find a time.

**Comment by kRITZCREEK** [comment]
The last commit changed the AST representation of `Abs` to always use a Binder instead of `Either Ident Binder`. The `Left Ident` is now represented as a `VarBinder`.

I'm sure at this point this PR has become a mess to merge after #2652 but if this seems good we could maybe try to incorporate this change into that PR. I checked and `pscid` still compiles fine with these changes.

File: src/Language/PureScript/TypeChecker/Types.hs [hunk 2]
```diff
@@ -636,11 +637,12 @@ check' (Literal (ArrayLiteral vals)) t@(TypeApp a ty) = do
   unifyTypes a tyArray
   array <- Literal . ArrayLiteral <$> forM vals (`check` ty)
   return $ TypedValue True array t
-check' (Abs (Left arg) ret) ty@(TypeApp (TypeApp t argTy) retTy) = do
-  unifyTypes t tyFunction
-  ret' <- withBindingGroupVisible $ bindLocalVariables [(arg, argTy, Defined)] $ check ret retTy
-  return $ TypedValue True (Abs (Left arg) ret') ty
-check' (Abs (Right _) _) _ = internalError "Binder was not desugared"
+check' (Abs binder ret) ty@(TypeApp (TypeApp t argTy) retTy)
+  | (VarBinder arg) <- binder = do
```
Comment (paf31) [inline src/Language/PureScript/TypeChecker/Types.hs]: Redundant parens here

File: src/Language/PureScript/TypeChecker/Types.hs [hunk 3]
```diff
@@ -348,12 +348,13 @@ infer' (Accessor prop val) = withErrorMessageHint (ErrorCheckingAccessor val pro
...
   typed <- check val (TypeApp tyRecord (RCons (Label prop) field rest))
   return $ TypedValue True (Accessor prop typed) field
-infer' (Abs (Left arg) ret) = do
-  ty <- freshType
-  withBindingGroupVisible $ bindLocalVariables [(arg, ty, Defined)] $ do
-    body@(TypedValue _ _ bodyTy) <- infer' ret
-    return $ TypedValue True (Abs (Left arg) body) $ function ty bodyTy
-infer' (Abs (Right _) _) = internalError "Binder was not desugared"
+infer' (Abs binder ret)
+  | (VarBinder arg) <- binder = do
```
Comment (paf31) [inline src/Language/PureScript/TypeChecker/Types.hs]: Same

File: src/Language/PureScript/Parser/Declarations.hs (hunk 1 above)
Comment (paf31) [inline src/Language/PureScript/Parser/Declarations.hs]: The PR seems good to me, but I'd really like to be able to remove this. Seems the parser should be able to guarantee we only add one `PositionedBinder`.

**Comment by kRITZCREEK** [comment]
I've applied your suggestions and removed the changes to the parser.

**Comment by paf31** [comment]
@kRITZCREEK Looks good, but I didn't want you to necessarily remove all of that code, just to hopefully find a way to remove the duplicate annotations. We can come back to this though.

**Comment by kRITZCREEK** [comment]
I still have all the code so it's no big deal. I'm hoping @garyb's PR will subsume all of it anyway.

**Comment by paf31** [comment]
Yes, that sounds like a good idea. Many libraries don't build right now because of the breaking changes in 0.11, but most core libraries do.

**Comment by kRITZCREEK** [comment]
I've readded the position information for binder parsers and got around the special check by making a desugaring step in the parser recurse on positioned binders properly. This compiles both pscid and slamdata fine except for discard errors in slamdata, but these are to be expected.

**Comment by paf31** [comment]
👍 Thanks, looks good to me!

**Comment by kRITZCREEK** [comment]
Cool I'll tidy the commit history up and merge then, @garyb is that OK with you? This will probably cause lots of conflicts with your AST PR.

**Comment by garyb** [comment]
Yeah, no problem. I might start that PR over and try and hammer it out in one day, since it's getting kinda stale now :wink:. Will see how bad the conflicts are first though.

**Pull request merged**